custom_random_forest.py

```diff
@@ -0,0 +1,143 @@
+        - X_train: feature array (numpy array)
+        - y_train: label array (numpy array)
+        - n_jobs: number of processes for parallel tree fitting (default is 1)
+
+        Returns:
+        - Trained random forest model
+
+        Steps:
+        1. Identifies the unique classes in the labels and stores them in self.classes_
+        2. Initializes empty lists for the trees (self.trees) and feature indices for each tree (self.feat_ids_by_tree)
+        3. Sets the value of max_features if not provided:
+            - If max_features is greater than the number of features in the data, raises a ValueError
+            - Determines the value of max_features based on the number of features in the data
+        4. Creates a pool of arguments for each tree in the forest
+        5. Fits each tree in the forest in parallel using ProcessPoolExecutor
+        6. Saves the trained trees and feature indices for each tree
+        7. Returns the trained random forest model
+        """
+        X, y = X_train, y_train
+        self.classes_ = sorted(np.unique(y))
+        self.trees = []
+        self.feat_ids_by_tree = []
+
+        # set max_features (if None):
+        if self.max_features > X.shape[1]:
+            raise ValueError('`max_features` is greater than dataset features number')
+        if self.max_features is None:
+            if X.shape[1] < 3:
+                self.max_features = 1
+            elif X.shape[1] < 6:
+                self.max_features = 2
+            else:
+                self.max_features = X.shape[1] // 3
+
+        args_pool = [_ for _ in zip(
+            [(X, y, self.max_features, self.max_depth, self.random_state)] * self.n_estimators,
+            range(self.n_estimators))]
+
+        with ProcessPoolExecutor(n_jobs) as pool:
+            dtrees_results = list(pool.map(self._fit_single_dtree, args_pool))
+        for res in dtrees_results:
+            self.trees.append(res[0])
+            self.feat_ids_by_tree.append(res[1])
+
+        return self
+
+    def _fit_single_dtree(self, single_dtree_args):
+        X, y, n_features, depth, rand_state = single_dtree_args[0]
+        rand_state_modifier = single_dtree_args[1]
+        np.random.seed(rand_state + rand_state_modifier)
+
+        single_feat_ids = np.random.choice(range(X.shape[1]), n_features, replace=False)
+
+        # bootstrap:
+        sample_ids = np.random.randint(0, X.shape[0], size=X.shape[0])
+        X_sampled = X[sample_ids][:, single_feat_ids]
+        y_sampled = y[sample_ids]
+
+        tree_estimator = DecisionTreeClassifier(max_depth=depth,
+                                                max_features=n_features,
+                                                random_state=rand_state)
+        tree_estimator.fit(X_sampled, y_sampled)
+
+        return tree_estimator, single_feat_ids
+
+    def _single_predict_proba(self, args):
+        X_test, tree, feat_ids = args
+        X_test_subset = X_test[:, feat_ids]
```

Comment:
```suggestion
    def _predict_proba_single_tree(self, args):
        X_test, tree, feat_ids = args
        X_test_subset = X_test[:, feat_ids]
```

Чтобы консистентно было